### PR TITLE
Implement periodic boundary conditions

### PR DESCRIPTION
**Pull request by marinlauber**
Implementation of periodic boundary conditions together with the convective exit.

There is an example [examples/TwoD_circle_periodicBC_convectiveBC.jl](file) that runs a y-periodic flow with a convective exit. The circle moves in the y-direction and crosses the boundary. This shows how to set up a periodic sdf.

I have added tests to check that the correct behavior is obtained for the boundary condition, for the different combinations of input.

I have not tested that the QUICK scheme implementation for periodic BC is correct, I just assumed it is based on @TzuYaoHuang simulations.

#### One caveat: there is no check that x-periodic flow and convective exit are not used at the same time, but because this will default to periodic, it should be fine.

### CODE REVIEW

File: src/Body.jl [hunk 1]
```diff
@@ -49,9 +49,10 @@ function measure!(a::Flow{N,T},body::AbstractBody;t=T(0),ϵ=1) where {N,T}
         end
     end
     @loop fill!(a.μ₀,a.μ₁,a.V,a.σᵥ,a.σ,I) over I ∈ inside(a.p)
+    BC!(a.μ₀,zeros(SVector{N,T}),a.exitBC,a.perdir;Dirichlet=false)  # add BC care for periodic & neu for μ₀
```
Comment (weymouth) [inline src/Body.jl:52]: I don't understand these comments...

File: src/Flow.jl [hunk 2]
```diff
@@ -4,6 +4,7 @@
 @fastmath quick(u,c,d) = median((5c+2d-u)/6,c,median(10c-9u,c,d))
 @fastmath vanLeer(u,c,d) = (c≤min(u,d) || c≥max(u,d)) ? c : c+(d-c)*(c-u)/(d-u)
 @inline ϕu(a,I,f,u,λ=quick) = @inbounds u>0 ? u*λ(f[I-2δ(a,I)],f[I-δ(a,I)],f[I]) : u*λ(f[I+δ(a,I)],f[I],f[I-δ(a,I)])
+@inline ϕuSelf(I1,I2,I3,I4,f,u,λ=quick) = @inbounds u>0 ? u*λ(f[I1],f[I2],f[I3]) : u*λ(f[I4],f[I3],f[I2])
```
Comment (weymouth) [inline src/Flow.jl:7]: need a few tests

File: src/Flow.jl [hunk 3]
```diff
@@ -32,18 +33,33 @@ function median(a,b,c)
...
+        upperBoundary!(r,u,Φ,ν,i,j,N,Val{tagper}())
     end
 end
 
+# Neumann BC Building block
+lowerBoundary!(r,u,Φ,ν,i,j,N,::Val{false}) = @loop r[I,i] += ϕuL(j,CI(I,i),u,ϕ(i,CI(I,j),u)) - ν*∂(j,CI(I,i),u) over I ∈ slice(N,2,j,2)
+upperBoundary!(r,u,Φ,ν,i,j,N,::Val{false}) = @loop r[I-δ(j,I),i] += -ϕuR(j,CI(I,i),u,ϕ(i,CI(I,j),u)) + ν*∂(j,CI(I,i),u) over I ∈ slice(N,N[j],j,2)
+
+# Periodic BC Building block
+lowerBoundary!(r,u,Φ,ν,i,j,N,::Val{true}) = @loop (
```
Comment (weymouth) [inline src/Flow.jl:58]: needs test

File: src/Flow.jl [hunk 4]
```diff
@@ -110,14 +128,14 @@ and the `AbstractPoisson` pressure solver to project the velocity onto an incomp
 @fastmath function mom_step!(a::Flow,b::AbstractPoisson)
     a.u⁰ .= a.u; scale_u!(a,0)
     # predictor u → u'
-    conv_diff!(a.f,a.u⁰,a.σ,ν=a.ν)
-    BDIM!(a); BC!(a.u,a.U,a.exitBC)
+    conv_diff!(a.f,a.u⁰,a.σ,ν=a.ν,perdir=a.perdir)
```
Comment (weymouth) [inline src/Flow.jl:131]: These function arguments are getting pretty long... It's fine, I guess

File: src/MultiLevelPoisson.jl [hunk 5]
```diff
@@ -74,14 +76,16 @@ function Vcycle!(ml::MultiLevelPoisson;l=1)
...
     prolongate!(fine.ϵ,coarse.x)
+    BC!(fine.ϵ;perdir=fine.perdir)
     increment!(fine)
 end
 
 mult!(ml::MultiLevelPoisson,x) = mult!(ml.levels[1],x)
 residual!(ml::MultiLevelPoisson,x) = residual!(ml.levels[1],x)
 
-function solver!(ml::MultiLevelPoisson;log=false,tol=1e-3,itmx=32)
+function solver!(ml::MultiLevelPoisson;log=false,tol=1e-6,itmx=64)
```
Comment (weymouth) [inline src/MultiLevelPoisson.jl:86]: Shouldn't change the default

File: src/Poisson.jl [hunk 6]
```diff
@@ -112,6 +117,7 @@ function pcg!(p::Poisson;it=6)
     end
 end
 smooth!(p) = pcg!(p)
+# smooth!(p) = get_backend(p.r)==CPU() ? SOR!(p,it=3) : Jacobi!(p,it=20)
```
Comment (weymouth) [inline src/Poisson.jl:120]: don't want this comment

File: src/WaterLily.jl [hunk 7]
```diff
@@ -23,6 +23,8 @@ export AutoBody,measure,sdf,+,-
 
 include("Metrics.jl")
 
+abstract type AbstractSimulation end
```
Comment (weymouth) [inline src/WaterLily.jl:26]: What's this?

File: src/util.jl [hunk 8]
```diff
@@ -32,7 +38,9 @@ end
 @inline inside_u(dims::NTuple{N}) where N = CartesianIndices((map(i->(2:i-1),dims)...,1:N))
 splitn(n) = Base.front(n),last(n)
 size_u(u) = splitn(size(u))
-
+function inside_uWB(dims::NTuple{N},j) where {N}
```
Comment (weymouth) [inline src/util.jl:41]: This seems to be unused...

File: src/util.jl [hunk 9]
```diff
@@ -126,7 +134,14 @@ Base.front(I::CartesianIndex) = CI(Base.front(I.I))
 
 Apply a vector function `f(i,x)` to the faces of a uniform staggered array `c`.
 """
-apply!(f,c) = @loop c[Ii] = f(last(Ii),loc(Ii)) over Ii ∈ CartesianIndices(c)
+apply!(f,c) = hasmethod(f,Tuple{Int,CartesianIndex}) ? applyV!(f,c) : applyS!(f,c)
+applyV!(f,c) = @loop c[Ii] = f(last(Ii),loc(Ii)) over Ii ∈ CartesianIndices(c)
+""" 
+    apply!(f, c)
```
Comment (weymouth) [inline src/util.jl:140]: I think you need to just write one function description for both function. Here you haven't explained the difference and it seems confusing.

File: src/util.jl [hunk 10]
```diff
@@ -146,37 +161,78 @@ condition `a[I,i]=A[i]` is applied to the vector component _normal_ to the domai
...
-function BC!(a,A,saveexit=false)
+function BC!(a,A,saveexit=false,perdir=(0,);Dirichlet=true)
     N,n = size_u(a)
-    for j ∈ 1:n, i ∈ 1:n
-        if i==j # Normal direction, Dirichlet
-            for s ∈ (1,2)
-                @loop a[I,i] = A[i] over I ∈ slice(N,s,j)
+    for i ∈ 1:n, j ∈ 1:n
+        if j in perdir
+            @loop a[I,i] = a[CIj(j,I,N[j]-1),i] over I ∈ slice(N,1,j)
```
Comment (weymouth) [inline src/util.jl:168]: Need tests

File: src/util.jl [hunk 11]
```diff
@@ -146,37 +161,78 @@ condition `a[I,i]=A[i]` is applied to the vector component _normal_ to the domai
...
 Apply zero Neumann boundary conditions to the ghost cells of a _scalar_ field.
 """
-function BC!(a)
+function BC!(a;perdir=(0,))
     N = size(a)
     for j ∈ eachindex(N)
-        @loop a[I] = a[I+δ(j,I)] over I ∈ slice(N,1,j)
-        @loop a[I] = a[I-δ(j,I)] over I ∈ slice(N,N[j],j)
+        if j in perdir
+            @loop a[I] = a[CIj(j,I,N[j]-1)] over I ∈ slice(N,1,j)
```
Comment (weymouth) [inline src/util.jl:203]: tests....

File: src/util.jl [hunk 12]
```diff
@@ -146,37 +161,78 @@ condition `a[I,i]=A[i]` is applied to the vector component _normal_ to the domai
...
+            @loop a[I] = a[CIj(j,I,2)] over I ∈ slice(N,N[j],j)
+        else
+            @loop a[I] = a[I+δ(j,I)] over I ∈ slice(N,1,j)
+            @loop a[I] = a[I-δ(j,I)] over I ∈ slice(N,N[j],j)
+        end
+    end
+end
+
+"""
+    interp(x::SVector, arr::AbstractArray)
```
Comment (weymouth) [inline src/util.jl:213]: Why is this function part of the PR?

File: test/runtests.jl [hunk 13]
```diff
@@ -56,11 +57,22 @@ arrays = setup_backends()
...
 
         @allowscalar u[end,:,1] .= 3
-        BC!(u, U, true) # save exit values
+        BC!(u,U,true) # save exit values
         @allowscalar @test all(u[end, :, 1] .== 3)
 
         WaterLily.exitBC!(u,u,U,0) # conservative exit check
         @allowscalar @test all(u[end,2:end-1, 1] .== U[1])
+
+        BC!(u,U,true,(2,)) # periodic in y and save exit values
```
Comment (weymouth) [inline test/runtests.jl:66]: Not sure why, but the code cov tool doesn't seem to think these tests are actually running the new code. Give it a look.

File: test/runtests.jl [hunk 14]
```diff
@@ -98,20 +110,20 @@ end
     @test_throws AssertionError("MultiLevelPoisson requires size=a2ⁿ, where n>2") Poisson_setup(MultiLevelPoisson,(15+2,3^4+2))
 
     err,pois = Poisson_setup(MultiLevelPoisson,(10,10))
-    @test pois.levels[3].D == Float32[0 0 0 0; 0 -2 -2 0; 0 -2 -2 0; 0 0 0 0]
+    @test pois.levels[3].D == Float32[0 0 0 0; 0 -2 -3 0; 0 -3 -4 0; 0 0 0 0]
```
Comment (weymouth) [inline test/runtests.jl:113]: Why change this?

**Review by weymouth — COMMENTED** [review]
Looks good overall. 
 - There are quite a few unused functions and features. Let's cut those out and add them as a separate PR if needed.
 - I'm not sure why, but codecov doesn't think these tests are working. The coverage is only showing at 65% of the new lines, which is way too low.

File: test/runtests.jl (hunk 14 above)
Comment (marinlauber) [reply test/runtests.jl:113]: This is actually a bad error, I didn't realise when I did that, but it's related the the BC on the \mu_0 field in the levels below 1 of the multigrid.

File: src/Flow.jl (hunk 2 above)
Comment (marinlauber) [reply src/Flow.jl:7]: This can actually be written in a neater way, there is actually only `I-2δ(a,I)` that needs a special treatment.

File: src/Flow.jl (hunk 4 above)
Comment (marinlauber) [reply src/Flow.jl:131]: That's true, this will be resolved by `@generated` functions further down the line I suppose.

File: test/runtests.jl (hunk 13 above)
Comment (marinlauber) [reply test/runtests.jl:66]: Yes, I saw that the codecov doesn't register a few lines

File: src/Body.jl (hunk 1 above)
Comment (marinlauber) [reply src/Body.jl]: I have improved them

**Comment by marinlauber** [comment]
These two commits should do the trick, I think I have addressed all the issues and I tried to improve the coverage of the tests.
